### PR TITLE
upload total_reads.json if present

### PR DESCRIPTION
**Pull request by boris-dimitrov**
needed for running post-filtered inputs

### CODE REVIEW

File: app/models/sample.rb [hunk 1]
```diff
@@ -179,11 +179,24 @@ def initiate_input_file_upload
...
     stderr_array = []
+    total_reads_json = "total_reads.json"
+    total_reads_json_path = nil
     input_files.each do |input_file|
       fastq = input_file.source
+      total_reads_json_path = File.join(File.dirname(fastq.to_s), total_reads_json)
       _stdout, stderr, status = Open3.capture3("aws", "s3", "cp", fastq.to_s, "#{sample_input_s3_path}/#{input_file.name}")
       stderr_array << stderr unless status.exitstatus.zero?
     end
+    if total_reads_json.present?
```
Comment (cdebourcy) [inline app/models/sample.rb]: did you mean `if total_reads_json_path.present?` ?

File: app/models/sample.rb [hunk 2]
```diff
@@ -179,11 +179,24 @@ def initiate_input_file_upload
...
+      # For samples where we are only given fastas post host filtering, we need to input the total reads (before host filtering) from this file.
+      _stdout, _stderr, status = Open3.capture3("aws", "s3", "cp", total_reads_json_path, "#{sample_input_s3_path}/#{total_reads_json}")
+      # We don't have a good way to know if this file should be present or not, so we just try to upload it
+      # and if it fails, we try one more time;  if that fails too, it's okay, the file is optional.
```
Comment (cdebourcy) [inline app/models/sample.rb]: You can check whether file is present as follows?
```
_stdout, _stderr, status = Open3.capture3("aws", "s3", "ls", total_reads_json_path)
unless status.exitstatus.zero?
  ...
```

File: app/models/sample.rb (hunk 1 above)
Comment (boris-dimitrov) [reply app/models/sample.rb]: Yes!

File: app/models/sample.rb (hunk 2 above)
Comment (boris-dimitrov) [reply app/models/sample.rb]: Why would I want to do that?   If the file is present, I've just wasted an operation.  If the file is absent, I get an error the same way I would if I try to copy it.   Also, either approach can be a false negative for absence due to operational issues.   For this reason I am actually removing check_file_s3_presence from the python pipeline in https://github.com/chanzuckerberg/idseq-pipeline/pull/174

File: app/models/sample.rb (hunk 1 above)
Comment (boris-dimitrov) [reply app/models/sample.rb]: Thanks for catching this :)

File: app/models/sample.rb (hunk 2 above)
Comment (boris-dimitrov) [reply app/models/sample.rb]: Btw while working on the pipeline check_s3_file_presence this weekend I looked up the s3 api for that, something called head-object, with which you can distinguish more easily operational errors from file absence.  But it's too much work to be worth it at this time, I think.

File: app/models/sample.rb [hunk 3]
```diff
@@ -179,11 +179,24 @@ def initiate_input_file_upload
   def initiate_s3_cp
     return unless status == STATUS_CREATED
     stderr_array = []
+    total_reads_json = "total_reads.json"
```
Comment (jameshowardwang) [inline app/models/sample.rb]: seems like convention is to put this into a frozen string

Comment (boris-dimitrov) [reply app/models/sample.rb]: thanks!  i often forget that

Comment (boris-dimitrov) [reply app/models/sample.rb]: so I moved it to the top as a frozen constant, and then I am not sure if I should do this
```
      total_reads_json_path = File.join(File.dirname(fastq.to_s), TOTAL_READS_JSON)
```
or add `.to_s` like so:
```
      total_reads_json_path = File.join(File.dirname(fastq.to_s), TOTAL_READS_JSON.to_s)
```
Definitely testing :)

Comment (boris-dimitrov) [reply app/models/sample.rb]: Btw another problem I had here with testing is that I can't figure out how to import File into the ruby shell `irb`.   I ended up just spinning up rails console as somehow it has that magically imported.

Comment (boris-dimitrov) [reply app/models/sample.rb]: The `.to_s` appears unnecessary.

**Comment by boris-dimitrov** [comment]
Tested & implemented James' suggestion to freeze the filename constant.

**Pull request merged**